flang/lib/Optimizer/Builder/Runtime/Command.cpp

```diff
@@ -0,0 +1,66 @@
+//===-- Command.cpp -- generate assignment runtime API calls ---------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/Optimizer/Builder/Runtime/Command.h"
+#include "flang/Optimizer/Builder/FIRBuilder.h"
+#include "flang/Optimizer/Builder/Runtime/RTBuilder.h"
+#include "flang/Optimizer/Builder/BoxValue.h"
+#include "flang/Runtime/command.h"
+#include "llvm/ADT/Optional.h"
+using namespace Fortran::runtime;
+
+mlir::Value fir::runtime::genCommandArgumentCount(fir::FirOpBuilder & builder, mlir::Location loc)
+{
+    auto argumentCountFunc = fir::runtime::getRuntimeFunc<mkRTKey(ArgumentCount)>(loc, builder);
+    return builder.create<fir::CallOp>(loc, argumentCountFunc).getResult(0);
+}
+
+// GET_COMMAND_ARGUMENT intrinsic is split between 2 functions in implementation; ArgumentValue and ArgumentLength.
+// So we handle each seperately.
+void fir::runtime::genGetCommandArgument(fir::FirOpBuilder & builder, mlir::Location loc, mlir::Value number,
+                           llvm::Optional<fir::CharBoxValue> valueBox, mlir::Value length,
+                           mlir::Value status, llvm::Optional<fir::CharBoxValue> errmsgBox)
+{
+    auto argumentValueFunc = fir::runtime::getRuntimeFunc<mkRTKey(ArgumentValue)>(loc, builder);
+    auto argumentLengthFunc = fir::runtime::getRuntimeFunc<mkRTKey(ArgumentLength)>(loc, builder);
+
+    auto processCharBox = [&](llvm::Optional<fir::CharBoxValue> arg, mlir::Value &value) {
+        if (arg.hasValue()) {
+            value = builder.createBox(loc, *arg);
+        } else {
+            value = builder.create<fir::AbsentOp>(loc, fir::BoxType::get(builder.getNoneType()));
+        }
+    };
+
+    mlir::Value value;
+    processCharBox(valueBox, value);
+
+    mlir::Value errmsg;
+    processCharBox(errmsgBox, errmsg);
+
+    llvm::SmallVector<mlir::Value> args = fir::runtime::createArguments(
+        builder, loc, argumentValueFunc.getType(), number, value, errmsg
+    );
+    mlir::Value result = builder.create<fir::CallOp>(loc, argumentValueFunc, args).getResult(0);
```

Comment:
I may be missing something, but I think if you have a program like this:
```
program p

  integer :: len
  call get_command_argument(1, length=len)
  print *, "len=", len
end program p
```
there will be a call to `ArgumentValue` to copy nothing.

We probably should have a test specifically to check this case!

There is potentially a quite a large list of of options here, where status and value, status and length, errmsg and length, etc are called. Not sure we need to test every permutation, but at least test errmsg only, status only as well. [Unless someone else has said "we don't need to test that"].

Reply:
Good point. Tried it when I was originally testing my lowering, but forgot to add to the unit test. Will do in next patch.

Reply:
Done.